From: Karel Zak <kzak@redhat.com>
Date: Thu, 20 Sep 2007 22:34:30 +0000 (+0200)
Subject: login: login segfaults on EOF (rh#298461)
X-Git-Url: http://git.kernel.org/?p=utils%2Futil-linux-ng%2Futil-linux-ng.git;a=commitdiff_plain;h=e797d83232802cf439b2ed893e784d3636357349

login: login segfaults on EOF (rh#298461)

Stupid bug in audit code:

  $ login
  login: ^D
  login: ^D
  Segmentation fault

Signed-off-by: Karel Zak <kzak@redhat.com>
---

diff --git a/login-utils/login.c b/login-utils/login.c
index e582779..1af8792 100644
--- a/login-utils/login.c
+++ b/login-utils/login.c
@@ -330,12 +330,12 @@ logaudit(const char *tty, const char *username, const char *hostname,
 	audit_fd = audit_open();
 	if (audit_fd == -1)
 		return;
-	if (!pwd)
+	if (!pwd && username)
 		pwd = getpwnam(username);
 	if (pwd)
 		snprintf(buf, sizeof(buf), "uid=%d", pwd->pw_uid);
 	else
-		snprintf(buf, sizeof(buf), "acct=%s", username);
+		snprintf(buf, sizeof(buf), "acct=%s", username ? username : "(unknown)");
 
 	audit_log_user_message(audit_fd, AUDIT_USER_LOGIN,
 		buf, hostname, NULL, tty, status);
